File: 2016/10/solution.js
const fs = require("fs");
const path = require("path");

const input = fs
  .readFileSync(path.join(__dirname, "input.txt"), "utf-8", (err, data) => {
    if (err) {
      console.log(err);
    } else {
      return data;
    }
  })
  .split("\n");

class Bot {
  constructor(giveLow, giveHigh) {
    this.lowBot = null;
    this.highBot = null;
    this.lowOutput = null;
    this.highOutput = null;
    this.values = [];
    if (giveLow[0] === "bot") {
      this.lowBot = giveLow[1];
    } else {
      this.lowOutput = giveLow[1];
    }
    if (giveHigh[0] === "bot") {
      this.highBot = giveHigh[1];
    } else {
      this.highOutput = giveHigh[1];
    }
  }
  execute(botMap, outputMap) {
    const res = [];
    if (this.values.length === 2) {
      let lowValue = this.values[0];
      let highValue = this.values[1];
      if (lowValue > highValue) {
        [lowValue, highValue] = [highValue, lowValue];
      }
      if (this.lowBot !== null) {
        botMap.get(this.lowBot).values.push(lowValue);
        if (botMap.get(this.lowBot).values.length === 2) {
          res.push(this.lowBot);
        }
      } else {
        outputMap.set(this.lowOutput, lowValue);
      }
      if (this.highBot !== null) {
        botMap.get(this.highBot).values.push(highValue);
        if (botMap.get(this.highBot).values.length === 2) {
          res.push(this.highBot);
        }
      } else {
        outputMap.set(this.highOutput, highValue);
      }
    }
    return res;
  }
}

const botMap = new Map();
const outputMap = new Map();
const re = /\w+ \d+/g;
const valuesToGiveOut = [];

for (const line of input) {
  const parts = line.match(re).map((a) => {
    a = a.split(" ");
    a[1] = Number(a[1]);
    return a;
  });
  if (parts.length === 3) {
    botMap.set(parts[0][1], new Bot(parts[1], parts[2]));
  } else {
    valuesToGiveOut.push([parts[0][1], parts[1][1]]);
  }
}

const botQueue = [];

for (const [value, bot] of valuesToGiveOut) {
  botMap.get(bot).values.push(value);
  if (botMap.get(bot).values.length === 2) {
    botQueue.push(bot);
  }
}

for (let i = 0; i < botQueue.length; i++) {
  const botsToQueue = botMap.get(botQueue[i]).execute(botMap, outputMap);
  botQueue.push(...botsToQueue);
}

for (const [n, bot] of botMap) {
  if (bot.values.includes(61) && bot.values.includes(17)) {
    console.log(n);
    break;
  }
}

let part2 = 1;
for (let i = 0; i < 3; i++) {
  part2 *= outputMap.get(i);
}

console.log(part2);

File: 2016/10/solution.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"runtime"
	"sort"
	"strconv"
	"strings"
)

func main() {
	_, filename, _, _ := runtime.Caller(0)
	dirname := filepath.Dir(filename)
	inputFilePath := filepath.Join(dirname, "input.txt")
	data, err := os.ReadFile(inputFilePath)
	if err != nil {
		panic(err)
	}
	puzzleInput := strings.Split(string(data), "\n")
	parseInput(puzzleInput)
	fmt.Println(part1())
	fmt.Println(part2())
}

var (
	botRegex   = regexp.MustCompile(`^bot (\d+) gives low to (\w+) (\d+) and high to (\w+) (\d+)$`)
	valueRegex = regexp.MustCompile(`^value (\d+) goes to bot (\d+)$`)
	botMap     = make(map[int]*Bot)
	outputMap  = make(map[int]int)
)

type Destination struct {
	kind string
	id   int
}

type Bot struct {
	id         int
	microchips []int
	lowDst     *Destination
	highDst    *Destination
	history    map[int]struct{}
}

func (b *Bot) AddMicrochip(microchip int) {
	b.microchips = append(b.microchips, microchip)
	b.history[microchip] = struct{}{}
	if len(b.microchips) == 2 {
		b.AssessMicrochips()
	}
}

func (b *Bot) AssessMicrochips() {
	low, high := b.microchips[0], b.microchips[1]
	if low > high {
		low, high = high, low
	}
	b.microchips = b.microchips[:0]
	if b.lowDst.kind == "bot" {
		botMap[b.lowDst.id].AddMicrochip(low)
	} else {
		outputMap[b.lowDst.id] = low
	}
	if b.highDst.kind == "bot" {
		botMap[b.highDst.id].AddMicrochip(high)
	} else {
		outputMap[b.highDst.id] = high
	}
}

func NewBot(line string) *Bot {
	match := botRegex.FindStringSubmatch(line)
	if match == nil {
		panic(fmt.Sprintf("Error matching bot regex with : %s", line))
	}
	botId, err := strconv.Atoi(match[1])
	if err != nil {
		panic(err)
	}
	lowId, err := strconv.Atoi(match[3])
	if err != nil {
		panic(err)
	}
	lowDst := Destination{match[2], lowId}
	highId, err := strconv.Atoi(match[5])
	if err != nil {
		panic(err)
	}
	highDst := Destination{match[4], highId}
	return &Bot{botId, []int{}, &lowDst, &highDst, map[int]struct{}{}}
}

func parseInput(puzzleInput []string) {
	sort.Strings(puzzleInput)
	for _, line := range puzzleInput {
		if line[0] == 'b' {
			bot := NewBot(line)
			botMap[bot.id] = bot
		} else {
			match := valueRegex.FindStringSubmatch(line)
			if match == nil {
				panic(fmt.Sprintf("Error matching value regex with : %s", line))
			}
			value, err := strconv.Atoi(match[1])
			if err != nil {
				panic(err)
			}
			botId, err := strconv.Atoi(match[2])
			if err != nil {
				panic(err)
			}
			if _, ok := botMap[botId]; !ok {
				panic(fmt.Sprintf("Cannot find bot %d", botId))
			}
			botMap[botId].AddMicrochip(value)
		}
	}
}

func part1() int {
	for _, bot := range botMap {
		if _, ok1 := bot.history[61]; ok1 {
			if _, ok2 := bot.history[17]; ok2 {
				return bot.id
			}
		}
	}
	return -1
}

func part2() int {
	return outputMap[0] * outputMap[1] * outputMap[2]
}
